kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,399 @@
+
+	case err != nil:
+		return nil, err
+	}
+
+	return newReadWriteBucket(b.tx, b.table, &id), nil
+}
+
+// DeleteNestedBucket deletes the nested bucket and its sub-buckets
+// pointed to by the passed key. All values in the bucket and sub-buckets
+// will be deleted as well.
+func (b *readWriteBucket) DeleteNestedBucket(key []byte) error {
+	// TODO shouldn't empty key return ErrBucketNameRequired ?
+	if len(key) == 0 {
+		return walletdb.ErrIncompatibleValue
+	}
+
+	result, err := b.tx.tx.ExecContext(
+		context.TODO(),
+		"DELETE FROM "+b.table+" WHERE "+parentSelector(b.id)+" AND key=$1 AND value IS NULL",
+		key,
+	)
+	if err != nil {
+		return err
+	}
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows == 0 {
+		return walletdb.ErrBucketNotFound
+	}
+
+	return nil
+}
+
+// Put updates the value for the passed key.
+// Returns ErrKeyRequred if te passed key is empty.
+func (b *readWriteBucket) Put(key, value []byte) error {
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	// Prevent NULL being written for an empty value slice.
+	if value == nil {
+		value = []byte{}
+	}
+
+	var query string
+	if b.id == nil {
+		query = "INSERT INTO " + b.table + " (key, value, parent_id) VALUES($1, $2, $3) ON CONFLICT (key) WHERE parent_id IS NULL DO UPDATE SET value=$2 WHERE " + b.table + ".value IS NOT NULL"
+	} else {
+		query = "INSERT INTO " + b.table + " (key, value, parent_id) VALUES($1, $2, $3) ON CONFLICT (key, parent_id) WHERE parent_id IS NOT NULL DO UPDATE SET value=$2 WHERE " + b.table + ".value IS NOT NULL"
+	}
+
+	result, err := b.tx.tx.ExecContext(
+		context.TODO(),
+		query,
+		key, value, b.id,
+	)
+	if err != nil {
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows != 1 {
+		return walletdb.ErrIncompatibleValue
+	}
+
+	return nil
+}
+
+// Delete deletes the key/value pointed to by the passed key.
+// Returns ErrKeyRequred if the passed key is empty.
+func (b *readWriteBucket) Delete(key []byte) error {
+	if key == nil {
+		return nil
+	}
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	_, err := b.tx.tx.ExecContext(
+		context.TODO(),
+		"DELETE FROM "+b.table+" WHERE key=$1 AND "+parentSelector(b.id)+" AND value IS NOT NULL",
+		key,
+	)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Check is key points to bucket if no rows affected?
+	return nil
+}
+
+// ReadWriteCursor returns a new read-write cursor for this bucket.
+func (b *readWriteBucket) ReadWriteCursor() walletdb.ReadWriteCursor {
+	return newReadWriteCursor(b)
+}
+
+// Tx returns the buckets transaction.
+func (b *readWriteBucket) Tx() walletdb.ReadWriteTx {
+	return b.tx
+}
+
+// NextSequence returns an autoincrementing sequence number for this bucket.
+// Note that this is not a thread safe function and as such it must not be used
+// for synchronization.
+func (b *readWriteBucket) NextSequence() (uint64, error) {
+	seq := b.Sequence() + 1
```

Comment:
Yes, that is possible. I can add an optimization for that.

Reply:
Looked at it briefly. It may be possible to do in one query, but it would need to distinguish between top level buckets with the sequence number in a separate table and sub-buckets that track their own sequence number. Unless you consider it a blocker, I'd rather not introduce another set of queries at this point.